src/sinks/datadog/logs/service.rs

```diff
@@ -178,36 +154,16 @@ impl Service<LogApiRequest> for LogApiService {
             .expect("building HTTP request failed unexpectedly");
 
         Box::pin(async move {
-            match client.call(http_request).in_current_span().await {
-                Ok(response) => {
-                    let status = response.status();
-                    // From https://docs.datadoghq.com/api/latest/logs/:
-                    //
-                    // The status codes answered by the HTTP API are:
-                    // 200: OK (v1)
-                    // 202: Accepted (v2)
-                    // 400: Bad request (likely an issue in the payload
-                    //      formatting)
-                    // 403: Permission issue (likely using an invalid API Key)
-                    // 413: Payload too large (batch is above 5MB uncompressed)
-                    // 5xx: Internal error, request should be retried after some
-                    //      time
-                    match status {
-                        StatusCode::BAD_REQUEST => Err(LogApiError::BadRequest),
-                        StatusCode::FORBIDDEN => Err(LogApiError::Forbidden),
-                        StatusCode::OK | StatusCode::ACCEPTED => Ok(LogApiResponse {
-                            event_status: EventStatus::Delivered,
-                            count,
-                            events_byte_size,
-                            raw_byte_size,
-                            protocol,
-                        }),
-                        StatusCode::PAYLOAD_TOO_LARGE => Err(LogApiError::PayloadTooLarge),
-                        _ => Err(LogApiError::ServerError),
-                    }
-                }
-                Err(error) => Err(LogApiError::HttpError { error }),
-            }
+            let result = client.call(http_request).in_current_span().await;
+            DatadogApiError::from_result(result)?;
```

Comment:
Admittedly, this is obviously the most concise way to check if it's an error but also, it's weird because it doesn't end up doing anything with the return of `from_result`... I wish `Result<T, E>` had something equivalent to `and_then` that gave you the full `Result<T, E>`, not just methods for either the `T` or `E`... then this could be composed in a much cleaner way.

Either way, not a blocker.

Reply:
I'll take Jesse's suggestion below for this , but keep it called "from_result" , because the other usage of the function is not mapping it and the "from" in "from_result" is based on the consistent input, whereas the caller may choose to map it or not.